### ISSUE TITLE
Code Type C needs to be tweaked

### ISSUE THREAD

**Issue by XxUnkn0wnxX** [bucanero/apollo-lib#20]
@bucanero 
I just noticed that code type C does not work like how SW has implemented it in your build.
this is for operations 0 & 8.

I noticed this when I was writing an experimental code for death stranding 
this is the code I'm just using to write the string `--TEST--`
```
8001000A D5988217
05F1C489 55290000
C8250029 00000071
A8000029 00000008
2D2D5445 53542D2D
```

now your's stops on the 37'th search result, however when running it via SW it stops on the 38th result.
which is intended.

I think you misunderstood how C works, for code type C when doing `C0` or `C8`

It doesn’t include the reference bytes its using as a search result.
what I mean by "reference bytes" I mean the bytes it is told to search with from the address.
so the 1st search result should match the 2nd set of bytes it finds not the 1st set that's included at the address of where it grabs the bytes that it needs to search with.

here is a test code I used for apollo & SW on a test save to show you the difference:
```
8001000A D5988217
05F1C489 55290000
C8010029 00000071
A8000029 00000008
2D2D5445 53542D2D
```
same code not modified, yours edits the 1st result which is wrong, SW edits the 2nd result which is what the 1st result is in this case.
[Code Type C DS Saves Test.zip](https://github.com/bucanero/apollo-lib/files/13063914/Code.Type.C.DS.Saves.Test.zip)

so for operations 0 & 8, set it so it ignores the result from the pointer or addresses it's grabbing the bytes to search with. (it should not match itself.)

Operators `C4` & `CC` are working as intended (the search ends at the address/pointer)

**Comment by bucanero** [comment]
hi @XxUnkn0wnxX , thanks for reporting and sharing some examples to compare 👍 

I'll try on my end and see if I can get Apollo to behave as expected. I'll share some test builds while working on it so you can validate and verify with more tests/examples.

**Comment by XxUnkn0wnxX** [comment]
i'm trying to find a fix my self actually ;)
I'm not good with C++ getting chat gpt it help me debug it..

i think I'm close though.. just testing that the code behaves as expected..

**Comment by XxUnkn0wnxX** [comment]
@bucanero ok the above code works as expected I've tested all the operators & they all seem fine.

**Comment by XxUnkn0wnxX** [comment]
ok all good all operators are working as expected with the above logic @bucanero 
feel free to re-write it, as I got this fix debugging it with chat-gpt so its not my code per say.

**Comment by bucanero** [comment]
@XxUnkn0wnxX thanks for sharing the source file

on my side, I was doing some research, and now I have a question about the `C4` / `CC` cases.
I found this note about the `C` code type:
https://github.com/bucanero/apollo-ps3/wiki/Game-Genie-code-format#code-type-c-search--pointer-command-2-use-bytes-from-offset-to-search-with

there, it says:
- 4 = Search Backwards from Address Given
- C = Offset from Pointer; Search Backwards from Address Given

could you check with the original Save Wizard app, if `C4/CC` are actually doing a "backwards" search or not?

since I'm fixing the C type, it would be nice to check and verify all cases.

Right now the code I did for C4/CC is doing "Search from 0x0 to Address Given" but wanted to confirm. Thanks for the help

**Comment by XxUnkn0wnxX** [comment]
> @XxUnkn0wnxX thanks for sharing the source file
> 
> on my side, I was doing some research, and now I have a question about the `C4` / `CC` cases. I found this note about the `C` code type: https://github.com/bucanero/apollo-ps3/wiki/Game-Genie-code-format#code-type-c-search--pointer-command-2-use-bytes-from-offset-to-search-with
> 
> there, it says:
> 
> * 4 = Search Backwards from Address Given
> * C = Offset from Pointer; Search Backwards from Address Given
> 
> could you check with the original Save Wizard app, if `C4/CC` are actually doing a "backwards" search or not?
> 
> since I'm fixing the C type, it would be nice to check and verify all cases.
> 
> Right now the code I did for C4/CC is doing "Search from 0x0 to Address Given" but wanted to confirm. Thanks for the help

I'm pretty sure it's `Search from 0x0 to Address Given` This was discussed in the SW discord server a while back I do remember Dynamite & skiller discussing this.

**Comment by XxUnkn0wnxX** [comment]
ya it's from `0x0`
I shared the same files via Discord
[SW C4-CC Test.zip](https://github.com/bucanero/apollo-lib/files/13077671/SW.C4-CC.Test.zip)

both cases wrote at `0x4969B0` if it where backwards it would have wrote at `0x49F210`

**Comment by bucanero** [comment]
@XxUnkn0wnxX , I have added the fix and created a PR #21 

the fix is based on your code so I assume everything should be working, but if you can please test the the auto-builds from the PR, here: https://github.com/bucanero/apollo-lib/actions/runs/6632061906?pr=21

you can download the "macos artifact" at the bottom and test it out. Thanks again!
